File: packages/commerce-types/src/platforms/stripe.ts
/**
 * Stripe → Warp model mappings. Stripe represents money in the smallest
 * currency unit (cents for USD, no decimals for JPY); the mapping converts to
 * and from Warp's decimal `Money` correctly per currency.
 */

import type { CurrencyCode, Money } from "../money.js";
import type { Commitment, Party, Value } from "../primitives.js";
import { commitmentId, individual, newCommitment, partyId, valueId } from "../primitives.js";
import type { CommitmentState } from "../states.js";
import { applyCommitmentPath } from "../transitions.js";

/** The synthetic party recorded as the actor of adapter-built history entries. */
const ADAPTER_ACTOR = partyId("system:stripe-adapter");

export type StripePaymentIntentStatus =
  | "requires_payment_method"
  | "requires_confirmation"
  | "processing"
  | "succeeded"
  | "canceled";

export interface StripePaymentIntent {
  id: string;
  amount: number; // smallest currency unit
  currency: string; // lowercase ISO code
  status: StripePaymentIntentStatus;
  customer?: string;
}

export interface StripeCustomer {
  id: string;
  email?: string;
}

export interface StripePrice {
  id: string;
  unit_amount: number; // smallest currency unit
  currency: string;
}

/**
 * Currencies with no minor unit — the Stripe amount is already the whole unit.
 * (Uppercase; `minorUnitFactor` normalizes case.)
 */
const ZERO_DECIMAL = new Set([
  "JPY", "KRW", "VND", "CLP", "ISK", "XAF", "XOF", "XPF", "BIF", "DJF",
  "GNF", "KMF", "MGA", "PYG", "RWF", "UGX", "VUV",
]);

/**
 * Currencies with THREE minor digits (millimes/fils) — the Stripe amount is the
 * whole unit × 1000, NOT × 100. TND is in this package's featured currency list;
 * treating it as two-decimal made every TND amount 10× wrong.
 */
const THREE_DECIMAL = new Set(["TND", "BHD", "KWD", "OMR", "JOD"]);

/**
 * The integer factor between a currency's whole unit and its Stripe minor unit:
 * 1 (zero-decimal), 1000 (three-decimal), or 100 (the default two-decimal case).
 */
function minorUnitFactor(currency: string): number {
  const c = currency.toUpperCase();
  if (ZERO_DECIMAL.has(c)) return 1;
  if (THREE_DECIMAL.has(c)) return 1000;
  return 100;
}

/** Convert a Stripe minor-unit amount into Warp `Money`. */
export function fromStripeAmount(amount: number, currency: string): Money {
  const code = currency.toUpperCase() as CurrencyCode;
  return { amount: amount / minorUnitFactor(currency), currency: code };
}

/**
 * Convert Warp `Money` into a Stripe `{ amount, currency }`. Rounds to an
 * integer after scaling so `1.5 * 1000` is exactly `1500`, never `1499.9999`.
 */
export function toStripeAmount(money: Money): { amount: number; currency: string } {
  const amount = Math.round(money.amount * minorUnitFactor(money.currency));
  return { amount, currency: money.currency.toLowerCase() };
}

function intentState(pi: StripePaymentIntent): CommitmentState {
  switch (pi.status) {
    case "requires_payment_method":
    case "requires_confirmation":
    case "processing":
      return { type: "Proposed" };
    case "succeeded":
      return { type: "Accepted" };
    case "canceled":
      return { type: "Cancelled", by: partyId("stripe"), reason: "canceled", at: new Date().toISOString() };
  }
}

export function fromStripePaymentIntent(pi: StripePaymentIntent): Commitment {
  const buyer = pi.customer ? partyId(pi.customer) : partyId("stripe_guest");
  // Fresh Draft → canonical path to the final state, so the commitment carries
  // a valid history (see transitions.ts / checkI4TemporalIntegrity).
  const draft: Commitment = {
    ...newCommitment(buyer, partyId("stripe_merchant")),
    id: commitmentId(pi.id),
    subject: {
      offered: [],
      requested: [
        {
          id: valueId(),
          form: { kind: "Money", money: fromStripeAmount(pi.amount, pi.currency) },
          quantity: 1,
          state: { type: "Available" },
        },
      ],
    },
  };
  return applyCommitmentPath(draft, intentState(pi), ADAPTER_ACTOR, "stripe-adapter");
}

export function fromStripeCustomer(customer: StripeCustomer): Party {
  // Stripe customer payloads carry no locale here; en/USD/US is an explicit
  // fallback — overwrite `.locale` when the caller knows the real locale.
  return individual(partyId(customer.id), { language: "en", currency: "USD", jurisdiction: "US" });
}

export function fromStripePrice(price: StripePrice): Value {
  return {
    id: valueId(price.id),
    form: { kind: "Money", money: fromStripeAmount(price.unit_amount, price.currency) },
    quantity: 1,
    state: { type: "Available" },
  };
}

File: packages/commerce-types-py/src/warp_commerce_types/platforms/stripe.py
"""Stripe -> Warp model mappings. Stripe represents money in the smallest
currency unit (cents for USD, no decimals for JPY, millimes x1000 for TND); the
mapping converts to and from Warp's decimal ``Money`` correctly per currency,
reusing the minor-unit logic in ``money`` (factor = 10 ** decimals, never /100).
"""
from __future__ import annotations

from typing import Any, Dict, Optional

from pydantic import BaseModel

from .._models import Commitment, CommitmentSubject, Money, Party, PartyLocale, Value
from ..money import minor_unit_factor
from ..primitives import commitment_id, individual, new_commitment, party_id, value_id
from ..transitions import apply_commitment_path

# The synthetic party recorded as the actor of adapter-built history entries.
_ADAPTER_ACTOR = party_id("system:stripe-adapter")


class StripePaymentIntent(BaseModel):
    id: str
    amount: int            # smallest currency unit
    currency: str          # lowercase ISO code
    status: str            # requires_payment_method | requires_confirmation | processing | succeeded | canceled
    customer: Optional[str] = None


class StripeCustomer(BaseModel):
    id: str
    email: Optional[str] = None


class StripePrice(BaseModel):
    id: str
    unit_amount: int       # smallest currency unit
    currency: str


def from_stripe_amount(amount: int, currency: str) -> Money:
    """Convert a Stripe minor-unit amount into Warp ``Money``."""
    return Money(amount=amount / minor_unit_factor(currency), currency=currency.upper())


def to_stripe_amount(money: Money) -> Dict[str, Any]:
    """Convert Warp ``Money`` into a Stripe ``{amount, currency}``. Rounds to an
    integer after scaling so ``1.5 * 1000`` is exactly ``1500``, never ``1499.9999``."""
    return {
        "amount": round(money.amount * minor_unit_factor(money.currency)),
        "currency": money.currency.lower(),
    }


def _intent_state(pi: StripePaymentIntent) -> Dict[str, Any]:
    if pi.status in ("requires_payment_method", "requires_confirmation", "processing"):
        return {"type": "Proposed"}
    if pi.status == "succeeded":
        return {"type": "Accepted"}
    # canceled
    from ..primitives import now

    return {"type": "Cancelled", "by": party_id("stripe"), "reason": "canceled", "at": now()}


def from_stripe_payment_intent(pi: StripePaymentIntent) -> Commitment:
    buyer = party_id(pi.customer) if pi.customer else party_id("stripe_guest")
    subject = CommitmentSubject.model_validate(
        {
            "offered": [],
            "requested": [
                {
                    "id": value_id(),
                    "form": {"kind": "Money", "money": from_stripe_amount(pi.amount, pi.currency)},
                    "quantity": 1,
                    "state": {"type": "Available"},
                }
            ],
        }
    )
    draft = new_commitment(buyer, party_id("stripe_merchant")).model_copy(
        update={"id": commitment_id(pi.id), "subject": subject}
    )
    return apply_commitment_path(draft, _intent_state(pi), _ADAPTER_ACTOR, "stripe-adapter")


def from_stripe_customer(customer: StripeCustomer) -> Party:
    return individual(
        party_id(customer.id),
        PartyLocale(language="en", currency="USD", jurisdiction="US"),
    )


def from_stripe_price(price: StripePrice) -> Value:
    return Value.model_validate(
        {
            "id": value_id(price.id),
            "form": {"kind": "Money", "money": from_stripe_amount(price.unit_amount, price.currency)},
            "quantity": 1,
            "state": {"type": "Available"},
        }
    )
